### PR TITLE
Add Tests to the Shortener Microservice Project

### LINKED ISSUE #37305: URL Shortener Microservice needs tests
I'm creating this as a place to discuss how to create tests for this project, so as not to dominate #37185

> @ojeytonwilliams For URL Shortener, I need to POST URL (google for example) to shorten it and
> it should return `{ "original_url": "www.google.com", "short_url": 1 }`. When I access `$.get(getUserInput(''url'') + ''/api/shorturl/1'')`, I should be redirected to google.com.
> 
> I will need to either 1) store the original_url as a variable OR 2) POST and then GET in one function. Is there a way to do the first one because that would be cleaner and more readable? And where does the first getUserInput get called from actually?
> 
> Here's an excerpt of my test, second test incomplete:
> 
> ```yaml
> tests:
>   - text: 'It should handle a URL as parameter and return shortened URL'
>     testString: 'getUserInput => $.post(getUserInput('url') + '/api/shorturl/new').then(data => assert.exists(data.short_url), xhr => { throw new Error(xhr.responseText)})'
>   - text: 'It should handle a shortened URL and redirect to original link'
>     testString: 'getUserInput => $.get(getUserInput(''url'') + ''/api/shorturl/3'').then()
> ```

### PR DESCRIPTION
**Pull request by abbathaw**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Closes #37305

This PR replaces #37588 which I had to close after a git rebase disaster (sorry 😢). Part of the learning process on rebasing with upstream and origins.

Apologies for the delay in doing the fixes to from the last comments, I completely forgot I had a hanging PR here.

### CODE REVIEW

**Comment by RandellDawson** [comment]
@abbathaw PR #39314 (currently in draft stage) will change the demo url from Glitch.me to using a Repl.co.  It appears the new version is up and running at https://url-shortener.freecodecamp.repl.co/

@scissorsneedfoodtoo Would it be best to go ahead and have this PR reflect the new demo project instead of referencing the Glitch version?

**Comment by scissorsneedfoodtoo** [comment]
@RandellDawson, yes, thank you for catching that.

@abbathaw, could you update this PR to test that the sample project at https://url-shortener.freecodecamp.repl.co/ can't be submitted?

**Comment by abbathaw** [comment]
@scissorsneedfoodtoo @RandellDawson I updated the link in the first test. 

I have a question though, this test (and maybe a couple of other tests I saw) are using this glitch link (for backend testing) `https://fcc-back-end-tester.glitch.me`, will this remain on glitch or will you move it to repl too?

**Comment by RandellDawson** [comment]
@abbathaw The test `text` should be written in the 2nd person (`You` not `I`).  See [Challenge description/instructions documentation](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/docs/how-to-work-on-coding-challenges.md#challenge-descriptionsinstructions)

File: curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md
Comment (RandellDawson) [inline curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md:27]:
```suggestion
You should provide your own project, not the example URL.
```

Comment (abbathaw) [reply curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md]: @RandellDawson  updated.

Comment (abbathaw) [reply curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md]: Maybe I should update the wording in the other available (already merged) tests in the category of apis-and-microservices such as `exercise-tracker`, "timestamp-microservice" file so they are consistent. The wording in the `request-header-parser-microservice`test looks ok. I can do that in another PR to not mix the scope of this one.

Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md]: @abbathaw, sure, that sounds like a great idea. Go ahead and update that text in another PR whenever you get a chance.

**Comment by scissorsneedfoodtoo** [comment]
@abbathaw, thank you for your patience and changing the regex check for the new project example's URL.

I reviewed this locally and everything works, but I saw the POST requests to `https://fcc-back-end-tester.glitch.me` that you pointed out. As far as I'm aware, that application was created by a contributor. While it seems to work really well for all the projects I tested it with, I think we'd like to do as much testing as possible without relying on a 3rd party app.

I played around with the tests locally and have some suggestions. There are one or two of downsides, though, which I'll include in my review comments.

File: curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md:80]: Here's the part where I'm not entirely sure if it's best practices or not:

```js
const getResponse = await fetch(url + '/api/shorturl/' + shortenedUrlVariable)
        .then(res => res)

      const getResponse = await fetch(url + '/api/shorturl/' + shortenedUrlVariable)
        .then(res => res)

      if (getResponse) {
        const { redirected, url } = getResponse;

        assert.isTrue(redirected);
        assert.strictEqual(url, `https://url-shortener.freecodecamp.repl.co/rand-${urlVariable}`); 
      } else {
        throw new Error(`${getResponse.status} ${getResponse.statusText}`);
      }
```

My thinking is that, while something like `...projectURL/api/shorturl/1598603478832` would redirect to `https://url-shortener.freecodecamp.repl.co/rand-1598603478832` which 404's, `fetch` still returns enough information for us to determine that the redirect was successful.

Could use more input on this, though. What do you think @RandellDawson and @moT01?

File: curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md:96]: Let's use `strictEqual` here instead:

```js
assert.strictEqual(error.toLowerCase(), 'invalid url'); 
```

**Comment by scissorsneedfoodtoo** [comment]
I believe that's everything I noticed. Could you hold off on making any changes before we get more feedback, @abbathaw?

File: curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md
Comment (abbathaw) [reply curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.english.md:36]: great valid point here.

**Review by naomi-lgbt — DISMISSED** [review]
The file that I am seeing in this PR is still using the `glitch` url for the example project..

That link should be replaced as well, correct? 
https://url-shortener.freecodecamp.repl.co/

**Comment by RandellDawson** [comment]
> The file that I am seeing in this PR is still using the `glitch` url for the example project..
> 
> That link should be replaced as well, correct?
> https://url-shortener.freecodecamp.repl.co/

As long as the link is not part of the lines changes being made, it will not go into master once merged.

**Comment by nhcarrigan** [comment]
I was not aware of that! Thanks for the info! 😁

**Comment by moT01** [comment]
To try and get this merged by the end of the day, I have added Kris's suggestions and tried to clarify some of the test text. 

[Here's my project](https://ash-airport.glitch.me) if anyone wants to use it for testing.

**Review by RandellDawson — CHANGES_REQUESTED** [review]
The 2nd test is doing more than just validating original_url and short_url are returned as keys of the returned JSON.  It is actually validating the expected values for their properties are also returned.  Not sure how to best word the `text`, but I believe it should at least should mention the correct values should returned for the keys. Also, instead of "keys", I think it should say "properties".

**Comment by scissorsneedfoodtoo** [comment]
That's true, thanks for catching that @RandellDawson.

I did a bit of reading and it seems like "property" refers to the property name (or key)/value pair. So maybe the text could read:

```
You can POST a URL to <code>/api/shorturl/new</code> and get a JSON response with the expected <code>original_url</code> and <code>short_url</code> properties.
```

Or maybe something this to be more explicit:

```
You can POST a URL to <code>/api/shorturl/new</code> and get a JSON response with the property names <code>original_url</code> and <code>short_url</code> and the expected values for each.
```

**Review by RandellDawson — DISMISSED** [review]

**Comment by Sky020** [comment]
@RandellDawson or @moT01 How are you testing this? Forgive me, if I am being silly with how I think this should work.

I am constantly failing the 2nd and 3rd tests, because they expect my URL shortener to `dns.lookup` some non-existent URL (`https://url-shortener.freecodecamp.repl.co/rand-1600184905001`). So, naturally, my project throws an error saying:
```
ERR:  Error: getaddrinfo ENOTFOUND url-shortener.freecodecamp.repl.co/rand-1600184905001
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:66:26) {
  errno: 'ENOTFOUND',
  code: 'ENOTFOUND',
  syscall: 'getaddrinfo',
  hostname: 'url-shortener.freecodecamp.repl.co/rand-1600184905001'
}
```

So, is my app not supposed to check that the URL exists? If so, then this hint should be removed from the boilerplate:
> HINT: to be sure that the submitted url points to a valid site you can use the function dns.lookup(host, cb) from the dns core module.

**Comment by moT01** [comment]
You're probably right, what's your project @Sky020?

**Comment by Sky020** [comment]
@moT01  https://repl.it/@Sky020/boilerplate-project-urlshortener

**Comment by scissorsneedfoodtoo** [comment]
Thanks for bringing this up @Sky020, we should probably make some changes based on this. Originally I tested this with an older project, but it may have passed the tests because of improper error handling more than anything.

Did a bit of refreshing on `dns.lookup` and it turns out that it just expects a hostname, so passing it `freecodecamp.org` works as expected, but not `freecodecamp.org/news`, or even `freecodecamp.org/`. I used the `request` module in my old project to make sure the page was valid, though that might be considered outside the scope of the project.

We could consider modifying the hint to mention that detail. Otherwise learners would have to look at the docs and other sources to find that out.

And to handle cases where learners check the whole URL rather than just the host with `dns.lookup`, we could modify the POST request to a URL that we can be reasonably sure is live:

```
... await fetch(url + '/api/shorturl/new/', {
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        body: `url=https://timestamp-microservice.freecodecamp.repl.co/api/timestamp/${urlVariable}`
      });
```

Then update the `assert` statements to the following:
* `assert.match(original_url, new RegExp(`timestamp-microservice.freecodecamp.repl.co/api/timestamp/${urlVariable}`));`
* `assert.strictEqual(url, `https://timestamp-microservice.freecodecamp.repl.co/api/timestamp/${urlVariable}`); `

I tried to work out a way to redirect/fetch a Google search result like https://www.google.com/search?query=1600184905001/1000, but that lead to CORS issues.

Not sure if this is the best way to handle it, though. What do you all think?

**Comment by Sky020** [comment]
@scissorsneedfoodtoo I would have thought the tests would not need to do anything more than:
```
await fetch(url + '/api/shorturl/new/', {
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        body: 'url=https://www.google.com'
      });
```
I get that we do not want campers to hardcode `const originalURL = 'google.com';`, but my understanding of what the tests are looking for is:
- Original URL is returned (if it is a valid/real/reachable location)
- `short_url !== originalURL && short_url.length() < originalURL.length()` Or similar
- A short URL is returned
- If `<short_url>` is passed as a parameter, and exists, the page will redirect to `www.google.com` or similar

**Comment by scissorsneedfoodtoo** [comment]
@Sky020, that's true, we could definitely just validate those first three and be pretty certain that all the requirements for this project are met.

The tricky part, at least in my testing, is verifying that passing the `short_url` as a parameter actually causes the page to redirect. It seems like with `fetch`, the default behavior is to follow the redirect and return a response from that page. But https://www.google.com has CORS settings that don't allow you to fetch data from the page, at least on the client side. 

Using the example project, if you send the POST request from your message, you'll get the following response:

```
{
  "original_url": "https://www.google.com",
  "short_url": 1
}
```

Then if you fetch https://url-shortener.freecodecamp.repl.co/api/shorturl/1, there'll be some CORS errors and the test will fail. There's an option for fetch not to follow the redirect, but it doesn't seem to show the URL that it would redirect to in that case, and we wouldn't be able to verify if the project would redirect to the correct URL.

So I'm not so sure what the best option here is. If there's some sort of open endpoint we can fetch through the redirect, that would work, too.

**Comment by Sky020** [comment]
>  But https://www.google.com has CORS settings that don't allow you to fetch data from the page, at least on the client side.

@scissorsneedfoodtoo Are there any `freecodecamp.org` endpoints we can use without worrying about sleeping/over-pinging the server?

**Comment by scissorsneedfoodtoo** [comment]
@Sky020, sorry for the late response. I'll have to double check, but now that we're moving the example projects to their own repo and serving them ourselves we should have a lot of options.

Hitting an endpoint like the timestamp microservice is good if we want to ensure that the URL being saved and redirected to is unique and hasn't been submitted before. If we're not worried about that, we could use a static page like https://freecatphotoapp.com/. We shouldn't run into any CORS issues with that page.

**Comment by raisedadead** [comment]
This is blocked by #39024

**Comment by RandellDawson** [comment]
Fixed merge conflict, and PR #39024 has been merged, so removing `status: blocked` label.

File: curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.md
Comment (Sky020) [inline curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.md:28]:
```suggestion
      assert(!(new RegExp('.*/url-shortener-microservice\\.freecodecamp\\.rocks')).test(url));
```

**Review by ShaunSHamilton — DISMISSED** [review]
I assume this needs to be updated.

**Review by ShaunSHamilton — DISMISSED** [review]
After discussing with Randell, and deciding the URL does not need to exist, I adjusted my project, and am happy with these changes, once the note about using `dns.lookup` is removed.

**Note:** There will not be backwards-compatibility with these changes.

**Comment by scissorsneedfoodtoo** [comment]
Thanks for taking care of the merge conflict and updating the first test string @Sky020. I just updated the wording so it's more in line with the rest of the test text.

Everything LGTM :+1: :+1:

**Review by ShaunSHamilton — APPROVED** [review]
This LGTM 👍 

@scissorsneedfoodtoo Thank you, for catching the text wording.

**Comment by scissorsneedfoodtoo** [comment]
No worries, @Sky020. I almost missed it myself. 

Thank you for all the hard work you've done here and in all the other places around the repo :slightly_smiling_face:

**Review by RandellDawson — APPROVED** [review]
LGTM

**Comment by scissorsneedfoodtoo** [comment]
Thanks @abbathaw for writing these tests and making all those updates. Also, thank you everyone for your review. Everything LGTM :+1: :+1:

File: curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.md
Comment (Sky020) [inline curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.md:38]: @scissorsneedfoodtoo Should this not be: 
```
url=https://timestamp-microservice.freecodecamp.rocks/api/timestamp/${urlVariable}
```

Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/05-apis-and-microservices/apis-and-microservices-projects/url-shortener-microservice.md:38]: Totally missed that. Thank you for catching this and submitting a fix, @Sky020. I'll review it now.

**Review by adigson — COMMENTED** [review]
kindly guide to the actual review passing the shortened url project on freeCodeCamp.
Thanks

**Pull request merged**
